File: strategy/generator/confirm/roc.py
from dataclasses import dataclass

from core.models.parameter import Parameter, StaticParameter
from core.models.source import SourceType

from .base import Confirm, ConfirmType


@dataclass(frozen=True)
class RocConfirm(Confirm):
    type: Confirm = ConfirmType.Roc
    source_type: Parameter = StaticParameter(SourceType.CLOSE)
    period: Parameter = StaticParameter(21.0)

File: ta_lib/strategies/confirm/src/roc.rs
use base::prelude::*;
use core::prelude::*;
use momentum::roc;
use timeseries::prelude::*;

pub struct RocConfirm {
    source_type: SourceType,
    period: usize,
}

impl RocConfirm {
    pub fn new(source_type: SourceType, period: f32) -> Self {
        Self {
            source_type,
            period: period as usize,
        }
    }
}

impl Confirm for RocConfirm {
    fn lookback(&self) -> usize {
        self.period
    }

    fn validate(&self, data: &OHLCVSeries) -> (Series<bool>, Series<bool>) {
        let roc = roc(&data.source(self.source_type), self.period);

        (roc.sgt(&ZERO_LINE), roc.slt(&ZERO_LINE))
    }
}
